src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.cpp

```diff
@@ -696,11 +696,59 @@ class ShaderCallDither : public ShaderPart
 public:
 	ShaderCallDither(const opengl::GLInfo & _glinfo)
 	{
-		if (!_glinfo.isGLES2 && config.generalEmulation.enableNoise != 0) {
-			m_part =
-				"  if (uColorDitherMode == 2) colorNoiseDither(snoise(), clampedColor.rgb);	\n"
-				"  if (uAlphaDitherMode == 2) alphaNoiseDither(snoise(), clampedColor.a);	\n"
+		if (!_glinfo.isGLES2) {
+			if (config.generalEmulation.ditheringMode >= 1) {
+				m_part =
+					"  if (uColorDitherMode == 2) {												\n"
+					"    colorNoiseDither(snoiseRGB(), clampedColor.rgb);						\n"
+					"    quantizeRGB(clampedColor.rgb);											\n"
+					"  }																		\n"
+					"  if (uAlphaDitherMode == 2) {												\n"
+					"    alphaNoiseDither(snoiseA(), clampedColor.a);							\n"
+					"    quantizeA(clampedColor.a);												\n"
+					"  }																		\n"
+				;
+			}
+			if (config.generalEmulation.ditheringMode >= 3) {
+				m_part +=
+					// Try to keep dithering visible even at higher resolutions
+					"  lowp float divider = 1.0 + step(3.0, uScreenScale.x);					\n"
+					"  mediump ivec2 position = ivec2(mod((gl_FragCoord.xy - 0.5) / divider,4.0));\n"
+					"  																			\n"
+					"  lowp mat4 bayer = mat4( 0.0, 0.75, 0.1875, 0.9375,						\n"
+					"                            0.5, 0.25, 0.6875, 0.4375,						\n"
+					"                            0.125, 0.875, 0.0625, 0.8125,					\n"
+					"                            0.625, 0.375, 0.5625, 0.3125);					\n"
+					"  lowp mat4 mSquare = mat4( 0.0, 0.6875, 0.75, 0.4375,						\n"
+					"                              0.875, 0.3125, 0.125, 0.5625, 				\n"
+					"                              0.1875, 0.5, 0.9375, 0.25,					\n"
+					"                              0.8125, 0.375, 0.0625, 0.625);				\n"
+					"  																			\n"
+					"  lowp float bayerThreshold = bayer[position.x][position.y];				\n"
+					"  lowp float mSquareThreshold = mSquare[position.x][position.y];			\n"
```

Comment:
We will have performance problems here. It is known that fetch from matrix by dynamically calculated indices is slow. I already had that problem when I implemented shader-based blending. This part need to be rewritten using the same approach, as for the blending.

Reply:
Two 4x4 textures and two texelFetch() calls would do the same thing. Or switch statements with 16 cases.

Reply:
> Two 4x4 textures and two texelFetch() calls would do the same thing.

Yes, it would be fastest, but I don't want to write special code for special textures.

> Or switch statements with 16 cases.

I suppose it will not be faster if you switch on dynamic value.

I made it as for blending, with matrix-vertex multiplication and dot, commit 2f1cda1654